### PR TITLE
BGP basic configurations

### PR DESCRIPTION
**Pull request by eduolivares**
This patch includes kustomizations required to deploy BGP at different levels:
- metallb
- nad
- control-plane
- edpm

### CODE REVIEW

File: dt/bgp/kustomization.yaml
Comment (abays) [inline dt/bgp/kustomization.yaml:24]: L21, 23 and 24 can be simplified to `../../lib/networking`

**Comment by eduolivares** [comment]
> Could we get various `.md` `README`s included in this PR using the same pattern followed for other DTs and VAs? i.e. [1] [2] [3]
> 
> [1] https://github.com/openstack-k8s-operators/architecture/blob/main/examples/dt/uni01alpha/README.md [2] https://github.com/openstack-k8s-operators/architecture/blob/main/examples/dt/uni01alpha/control-plane.md [3] https://github.com/openstack-k8s-operators/architecture/blob/main/examples/dt/uni01alpha/data-plane.md

yes, I'll update the PR with them

File: dt/bgp/kustomization.yaml
Comment (abays) [reply dt/bgp/kustomization.yaml]: Sorry, I was completely wrong about this given the current way `lib/networking` is implemented.  It needs to be reversed to what you had beforehand.

Comment (eduolivares) [reply dt/bgp/kustomization.yaml:24]: ok, thanks. I'm running a d/s job to verify it. I'll let you know how it goes.
(I still have to add the README files)

Comment (abays) [reply dt/bgp/kustomization.yaml:24]: FYI: now either way works, as this [1] has been merged.

[1] https://github.com/openstack-k8s-operators/architecture/pull/196

Comment (eduolivares) [reply dt/bgp/kustomization.yaml]: ok, rebased and L21, L23 and L24 replaced by just `../../lib/networking`

I will trigger a d/s job with this change tonight.

Before this change, the PR was tested with d/s jobs and it worked fine (i.e. deployment finished successfully and the setup seems healthy, although we have some connectivity problems between test-operator and provider network IPs, but I don't think changes in this PR would fix it).
Hence, if we get similar results after this change, IMO we could merge this PR (when reviewers approve it, obviously).

**Comment by eduolivares** [comment]
> Could we get various `.md` `README`s included in this PR using the same pattern followed for other DTs and VAs? i.e. [1] [2] [3]
> 
> [1] https://github.com/openstack-k8s-operators/architecture/blob/main/examples/dt/uni01alpha/README.md [2] https://github.com/openstack-k8s-operators/architecture/blob/main/examples/dt/uni01alpha/control-plane.md [3] https://github.com/openstack-k8s-operators/architecture/blob/main/examples/dt/uni01alpha/data-plane.md

md files added.
@abays, please let me know how they look to you.

**Comment by abays** [comment]
> > Could we get various `.md` `README`s included in this PR using the same pattern followed for other DTs and VAs? i.e. [1] [2] [3]
> > [1] https://github.com/openstack-k8s-operators/architecture/blob/main/examples/dt/uni01alpha/README.md [2] https://github.com/openstack-k8s-operators/architecture/blob/main/examples/dt/uni01alpha/control-plane.md [3] https://github.com/openstack-k8s-operators/architecture/blob/main/examples/dt/uni01alpha/data-plane.md
> 
> md files added. @abays, please let me know how they look to you.

The MD files look good

**Comment by fultonj** [comment]
@eduolivares  This looks it's had some work behind it. Have you tested it or done any of  the things from this doc?

https://github.com/openstack-k8s-operators/architecture/pull/207

**Comment by eduolivares** [comment]
> @eduolivares This looks it's had some work behind it. Have you tested it or done any of the things from this doc?
> 
> #207

Yes. I tested this PR last week, using downstream jobs and it worked fine. Controlplane and Dataplane are successfully deployed.
I have some problems to run tests from the test-operator, but I think it's reasonable to merge this PR and resolve those issue later.
The reason is that I am not sure yet whether those problems will be fixed with changes in architecture (maybe metallb, but most probably not), test-operator, ci-fmw or some playbook in the downstream job.

**Comment by leifmadsen** [comment]
> > @eduolivares This looks it's had some work behind it. Have you tested it or done any of the things from this doc?
> > #207
> 
> Yes. I tested this PR last week, using downstream jobs and it worked fine. Controlplane and Dataplane are successfully deployed. I have some problems to run tests from the test-operator, but I think it's reasonable to merge this PR and resolve those issue later. The reason is that I am not sure yet whether those problems will be fixed with changes in architecture (maybe metallb, but most probably not), test-operator, ci-fmw or some playbook in the downstream job.

Awesome thanks! I'm satisfied with this PR since it contains automation changes, has been noted as being manually tested, and has self-contained changes that wouldn't affect base components (and thus VAs).

**Review by leifmadsen — APPROVED** [review]

**Comment by fultonj** [comment]
Would you please rebase this PR on the main branch?

Mainly you're missing this: https://github.com/openstack-k8s-operators/architecture/pull/197

When I test your PR using the current VA1 commands [1] it fails to build.


[1] https://github.com/openstack-k8s-operators/architecture/blob/main/examples/va/hci/dataplane-pre-ceph.md#initialize-pre-ceph

```
$ kustomize build edpm-pre-ceph/nodeset
$ kustomize build edpm-pre-ceph/deployment
$ kustomize build deployment
```

**Comment by leifmadsen** [comment]
> Would you please rebase this PR on the main branch?
> 
> Mainly you're missing this: #197

Huh I'm surprised we don't get a button for that here. Maybe we need to enable the feature in GitHub that requests PRs be up to date against `main`. When we enable that, there will be a button to allow us to automatically rebase these in the UI.

(I don't have admin permissions for this repo, so I can't enable it, or I would :))

File: examples/dt/bgp/dataplane.md
Comment (fultonj) [inline examples/dt/bgp/dataplane.md:26]: After rebasing on the main branch, this command will need to be updated since it will fail as below. Please split the dataplane nodeset and deployment into separate CR files.

```
[fultonj@runcible edpm{bgp-basic-config}]$ kustomize build > dataplane.yaml
Error: accumulating components: accumulateDirectory: "recursed accumulation of path '/home/fultonj/review/architecture/dt/bgp/edpm': accumulating components: accumulateDirectory: \"recursed accumulation of path '/home/fultonj/review/architecture/lib/dataplane': accumulating components: accumulateDirectory: \\\"recursed accumulation of path '/home/fultonj/review/architecture/lib/dataplane/nodeset': nothing selected by ConfigMap.[noVer].[noGrp]/edpm-nodeset-values.[noNs]:data.nodeset.ansible\\\"\""
[fultonj@runcible edpm{bgp-basic-config}]$
```

**Comment by fultonj** [comment]
> > Would you please rebase this PR on the main branch?
> > Mainly you're missing this: #197
> 
> Huh I'm surprised we don't get a button for that here. Maybe we need to enable the feature in GitHub that requests PRs be up to date against `main`. When we enable that, there will be a button to allow us to automatically rebase these in the UI.
> 
> (I don't have admin permissions for this repo, so I can't enable it, or I would :))

I do not have the Settings button to enable it either as described here.

https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-rebasing-for-pull-requests

Rebasing via the CLI will work though (I always do it that way myself in order to resolve conflicts).

**Comment by leifmadsen** [comment]
> > Huh I'm surprised we don't get a button for that here. Maybe we need to enable the feature in GitHub that requests PRs be up to date against `main`. When we enable that, there will be a button to allow us to automatically rebase these in the UI.
> > (I don't have admin permissions for this repo, so I can't enable it, or I would :))
> 
> I do not have the Settings button to enable it either as described here.
> 
> https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-rebasing-for-pull-requests
> 
> Rebasing via the CLI will work though (I always do it that way myself in order to resolve conflicts).

Yea that works fine. It was nice with STF repos so we could just see when the PR was out of sync with a visual indicator :)

**Comment by abays** [comment]
> > > Huh I'm surprised we don't get a button for that here. Maybe we need to enable the feature in GitHub that requests PRs be up to date against `main`. When we enable that, there will be a button to allow us to automatically rebase these in the UI.
> > > (I don't have admin permissions for this repo, so I can't enable it, or I would :))
> > 
> > 
> > I do not have the Settings button to enable it either as described here.
> > https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-rebasing-for-pull-requests
> > Rebasing via the CLI will work though (I always do it that way myself in order to resolve conflicts).
> 
> Yea that works fine. It was nice with STF repos so we could just see when the PR was out of sync with a visual indicator :)

We actually already have `Allow rebase merging` checked.  Do we want the option under it enabled?

image

**Comment by leifmadsen** [comment]
@abays yea that's the option I was referring to :)

**Comment by abays** [comment]
> @abays yea that's the option I was referring to :)

Cool.  I enabled it now.

**Comment by leifmadsen** [comment]
FWIW I'm working on https://github.com/openstack-k8s-operators/architecture/pull/193#discussion_r1585392698 as I hit the rebase button and broke this PR, so I'll submit another PR targeting this one with the expected changes.

File: examples/dt/bgp/dataplane.md
Comment (leifmadsen) [reply examples/dt/bgp/dataplane.md:26]: I have provided a PR against the base repo/branch to implement the refactoring. Please give it a check if you have a chance:  https://github.com/eduolivares/architecture/pull/1

**Comment by leifmadsen** [comment]
I opened https://github.com/eduolivares/architecture/pull/1 against the base repo/branch to fix the test-kustomize failure I caused when I did the PR rebase against main.

**Comment by eduolivares** [comment]
Running a d/s job to test @leifmadsen's dataplane split.

**Comment by eduolivares** [comment]
OK, several issues have been fixed since last week:
- dataplane split
- nncp default routes removed
- octavia configuration was incomplete (we didn't see it before because the octavia operator missed some validations)

Current version works. It has just been tested d/s and the deployment works. The tempest tests fail until we find a solution for [1], which is my next task.

This PR is ready for reviews again and hopefully ready for merge.

[1] https://issues.redhat.com/browse/OSPRH-4826

File: automation/vars/default.yaml
Comment (abays) [inline automation/vars/default.yaml:250]: This should maybe just be like this [1]?

[1] https://github.com/openstack-k8s-operators/architecture/blob/3e0b37a46e4bb47f1fba2b0bcf2bc2662b956d65/automation/vars/uni01alpha.yaml#L39

File: examples/dt/bgp/edpm/nodeset/kustomization.yaml
Comment (abays) [inline examples/dt/bgp/edpm/nodeset/kustomization.yaml:7]:
```suggestion
  ## It's possible to replace ../../../../../dt/bgp/edpm/nodeset/ with a git checkout URL as per:
```

File: dt/bgp/edpm/deployment/nova_custom.yaml
Comment (abays) [inline dt/bgp/edpm/deployment/nova_custom.yaml:1]: It might be better to put this in `nodeset` instead of `deployment`, because the other (normal) `OpenStackDataPlaneService`s are created when the `OpenStackDataPlaneNodeSet` is created [1][2].

[1] https://github.com/openstack-k8s-operators/dataplane-operator/blob/a1ed2be65e1f1ba6569afde0cea828d1ae8b3bb2/controllers/openstackdataplanenodeset_controller.go#L220-L230
[2] https://github.com/openstack-k8s-operators/dataplane-operator/blob/a1ed2be65e1f1ba6569afde0cea828d1ae8b3bb2/pkg/deployment/service.go#L75-L76

File: dt/bgp/edpm/deployment/kustomization.yaml
Comment (abays) [inline dt/bgp/edpm/deployment/kustomization.yaml:24]: See this comment: https://github.com/openstack-k8s-operators/architecture/pull/193/files#r1590951076

File: dt/bgp/edpm/deployment/nova_custom.yaml
Comment (abays) [inline dt/bgp/edpm/deployment/nova_custom.yaml:1]: It might be better to put this in `nodeset` instead of `deployment`, because the other (normal) `OpenStackDataPlaneService`s are created when the `OpenStackDataPlaneNodeSet` is created [1][2].

[1] https://github.com/openstack-k8s-operators/dataplane-operator/blob/a1ed2be65e1f1ba6569afde0cea828d1ae8b3bb2/controllers/openstackdataplanenodeset_controller.go#L220-L230
[2] https://github.com/openstack-k8s-operators/dataplane-operator/blob/a1ed2be65e1f1ba6569afde0cea828d1ae8b3bb2/pkg/deployment/service.go#L75-L76

**Review by abays — APPROVED** [review]
/lgtm

**Pull request merged**
